src/compute-client/src/types/dataflows.rs

```diff
@@ -67,17 +67,18 @@ pub struct DataflowDescription<P, S: 'static = (), T = mz_repr::Timestamp> {
 
 impl<T> DataflowDescription<Plan<T>, (), mz_repr::Timestamp> {
     /// Tests if the dataflow refers to a single timestamp, namely
-    /// that both `as_of` and `until` have a single coordinate and that
-    /// the `until` value corresponds to the `as_of` value plus one.
+    /// that `as_of` has a single coordinate and that the `until`
+    /// value corresponds to the `as_of` value plus one.
     pub fn is_single_time(&self) -> bool {
         // TODO: this would be much easier to check if `until` was a strict lower bound,
         // and we would be testing that `until == as_of`.
-        match (self.as_of.as_ref(), self.until.as_option()) {
-            (Some(as_of), Some(until)) => {
-                as_of.as_option().and_then(|as_of| as_of.checked_add(1)) == Some(*until)
-            }
-            _ => false,
-        }
+        let Some(as_of) = self.as_of.as_ref() else { return false; };
+        !as_of.is_empty()
+            && as_of
+                .as_option()
+                .and_then(|as_of| as_of.checked_add(1))
+                .as_ref()
+                == self.until.as_option()
```

Comment:
Is the `!as_of.is_empty()` check correct? If `as_of` is empty that means we will always ever only flow as single time through the dataflow, so shouldn't we return `true` in that case?

Assuming my reasoning checks out, I'd like to propose this version for improved readability:
```rust
        let Some(as_of) = self.as_of.as_ref() else { return false };
        let as_of_plus_one = as_of.as_option().and_then(|t| t.checked_add(1));
        as_of_plus_one.as_ref() == self.until.as_option()
```

Reply:
Recording here a discussion through another channel: A pTVC is defined in the interval of `[since, upper)`. My understanding when writing the condition was that when we spin a dataflow, we’d like it to process pTVCs that are valid in the interval `[as_of, until)`. So if the `as_of` is empty (i.e., the top element of the frontier poset), we can only have an interval with no times at all, not a single time.

This understanding can be checked by looking at the `as_of` that is picked when we query a constant collection:
```sql
materialize=> explain timestamp for select * from (values(1));
                       Timestamp                       
-------------------------------------------------------
                 query timestamp: 18446744073709551615+
 largest not in advance of upper: 18446744073709551615+
                           upper:[]                   +
                           since:[            0]      +
         can respond immediately: true                +
                        timeline: None                +
 
(1 row)

Time: 9.593 ms
```
We see that `u64::MAX` is selected for the query timestamp, not empty, matching the understanding above.